google/cloud/storage/internal/signed_url_requests.cc

```diff
@@ -185,6 +191,35 @@ std::string V4SignUrlRequest::StringToSign(std::string const& client_id) const {
          "\n" + Scope() + "\n" + CanonicalRequestHash(client_id);
 }
 
+Status V4SignUrlRequest::Validate() {
+  if (!virtual_host_name_) {
+    return Status();
+  }
+  auto const& headers = common_request_.extension_headers();
+  auto host_it = headers.find("host");
+  if (host_it == headers.end()) {
+    return Status();
+  }
+  return host_it->second == Hostname()
+             ? Status()
+             : Status(StatusCode::kInvalidArgument,
+                      "specified 'host' (" + host_it->second +
+                          ") header stands in conflict with "
+                          "'VirtualHostname' option.");
+}
+
+std::string V4SignUrlRequest::Hostname() {
+  return virtual_host_name_
+             ? (common_request_.bucket_name() + ".storage.googleapis.com")
+             : std::string("storage.googleapis.com");
```

Comment:
issue #3329 emphasizes on compatibility with S3. 
I think, setting up host name like this will cause interoperability issues.
What do you think?

Reply:
> I think, setting up host name like this will cause interoperability issues.

Can you clarify?  What incompatibility are you thinking about?

Reply:
as we do not allow any other `hostname`, we cannot set it something like `http://s3.amazonaws.com/bucket/`.

Although, I am not sure if this comes under interoperability.